posts/2024-06-05-mixing-python-envs.md

```diff
@@ -0,0 +1,181 @@
+environment:
+ * Is the package available from `conda-forge` or some other reputable conda channel?
+ * Is the package in line with the scope of the environment (i.e. climate data analysis)?
+ * Is the package actively maintained?
+ * Is the package and its dependencies compatible with the existing environment?
+
+If the answer to all of these questions is yes, then CMS will install the package into the environment on request. If, however, these conditions aren't met, that doesn't mean that you need to create a new conda environment
+from scratch to use a new package. 
+
+## `pip install --user`
+
+Lets use the [xWMB](https://github.com/hdrake/xwmb) package as an example.
+There is no conda distribution for this package, it can only be installed by `pip` from a git repository so it does not meet our criteria above. This
+package also has an extensive list of dependencies, most of which are available in the `conda/analysis3` environment, so using the `analysis3` 
+environment as a base will save quite a lot of disk space.
+ 
+You may be tempted to load the `conda/analysis3` environment and `pip install --user` the xWMB package.
+<terminal-window typingDelay=30 lineDelay=200>
+    <terminal-line data="input">module use /g/data/hh5/public/modules</terminal-line>
+    <terminal-line data="input">module load conda/analysis3</terminal-line>
+    <terminal-line data="input">pip install -\\\-user git+https://github.com/hdrake/xwmb.git@main</terminal-line>
+    <terminal-line>Collecting git+https://github.com/hdrake/xwmb.git@main</terminal-line>
+    <terminal-line>  Cloning https://github.com/hdrake/xwmb.git (to revision main) to /scratch/v45/dr4292/tmp/pip-req-build-qox565pe</terminal-line>
+    <terminal-line>  Running command git clone -\\\-filter=blob:none -\\\-quiet https://github.com/hdrake/xwmb.git /scratch/v45/dr4292/tmp/pip-req-build-qox565pe</terminal-line>
+    <terminal-line>  ...</terminal-line>
+    <terminal-line>Successfully installed regionate-0.0.1 sectionate-0.2.1 xbudget-0.1.0 xgcm-0.8.2.dev15+g7492277 xwmb-0.1.1 xwmt-0.1.1</terminal-line>
+</terminal-window>
+
+The trouble with this is that it can lead to problems in the future. The `analysis3` module
+is updated to every 3 months as new environments are released. For example, if you load `conda/analysis3` today,
+you'll get the `conda/analysis3-24.01` environment. If you were to load the same module a couple of months from
+now, you'll get the `conda/analysis3-24.04` environment. Our policies around updating the conda environments
+can be found [here](https://climate-cms.org/cms-wiki/services/services-conda.html). As time goes on, your
+locally installed package will get further and further out of date, and could potentially conflict with newer
+packages installed in updated analysis environments. 
+
+This is one of the most common issues we see come into our helpdesk. A user will open a JupyterLab session in ARE
```

Comment:
Like above, I don't think this creates conflict as the package wouldn't re-install xarray etc . It does if people have installed it without loading our envs first. Also if you use export ... you can easily fix this, as you're pointing out later.
Don't get me wrong if you want to re-use this package regularly then clearly creating a virtual env is best. I just don't want people who might feel daunted by creating their own env to give up (and yes they are out there)